src/v/archival/archival_policy.h

```diff
@@ -44,6 +61,22 @@ struct upload_candidate_with_locks {
     std::vector<ss::rwlock::holder> read_locks;
 };
 
+/// Wraps an error with an offset range, so that no
+/// further upload candidates are created from this offset range.
+struct skip_offset_range {
+    model::offset begin_offset;
+    model::offset end_offset;
+    candidate_creation_error reason;
+
+    friend std::ostream& operator<<(std::ostream&, const skip_offset_range&);
+};
+
+using candidate_creation_result = std::variant<
+  std::monostate,
```

Comment:
looks like `upload_candidate_with_locks` default constructible, if it is, then why is monostate needed?

Reply:
curious about this

Reply:
`monostate` is useful here because of the way the type is used here https://github.com/redpanda-data/redpanda/blob/dev/src/v/archival/ntp_archiver_service.cc#L1556 in a visitor.

`monostate` forms one branch of the visit and `upload_candidate_with_locks` forms another distinct branch. If we relied on default construction we would have to add a check in the `upload_candidate_with_locks` branch to figure out if we were looking at the default constructed object. Perhaps `upload_candidate_with_locks` should not be default constructable.

Reply:
thanks @abhijat